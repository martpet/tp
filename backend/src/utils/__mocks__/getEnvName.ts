import { EnvName } from '~/types';

const envName: EnvName = 'production';

export const getEnvName = vi.fn().mockName('getEnvName').mockReturnValue(envName);
