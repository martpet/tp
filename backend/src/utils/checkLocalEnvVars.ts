import { EnvName } from '~/types';

export const checkLocalEnvVars = (envName: EnvName) => {
  const names = [
    'VITE_PERSONAL_SUBDOMAIN',
    'PERSONAL_DEV_ACCOUNT_ID',
    'PERSONAL_DEV_ACCOUNT_REGION',
  ];
  if (envName === 'personal') {
    names.forEach((varName) => {
      if (!process.env[varName]) {
        throw new Error(`Environment variable "${varName}" is missing from .env.local`);
      }
    });
  }
};
