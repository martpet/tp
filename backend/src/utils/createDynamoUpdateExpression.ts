export const createDynamoUpdateExpression = (data: Record<string, unknown>) => {
  const ExpressionAttributeNames: Record<string, string> = {};
  const ExpressionAttributeValues: Record<string, unknown> = {};
  const actions: string[] = [];

  Object.entries(data).forEach(([attr, val]) => {
    actions.push(`#${attr} = :${attr}`);
    ExpressionAttributeNames[`#${attr}`] = attr;
    ExpressionAttributeValues[`:${attr}`] = val;
  });

  return {
    UpdateExpression: `set ${actions.join(',')}`,
    ExpressionAttributeNames,
    ExpressionAttributeValues,
  };
};
