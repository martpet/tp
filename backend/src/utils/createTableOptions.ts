import { RemovalPolicy } from 'aws-cdk-lib';
import {
  AttributeType,
  BillingMode,
  GlobalSecondaryIndexProps,
  TableProps,
} from 'aws-cdk-lib/aws-dynamodb';
import deepmerge from 'deepmerge';
import { Merge } from 'type-fest';

import { appName } from '~/consts/appConsts';

type Attr<T> = {
  [K in keyof T]?: {
    name: K;
    type: T[K] extends string
      ? AttributeType.STRING
      : T[K] extends number
      ? AttributeType.NUMBER
      : never;
  };
}[keyof T];

type Props<T> = Merge<
  TableProps,
  {
    tableName: string;
    partitionKey: Attr<T>;
    timeToLiveAttribute?: keyof T;
    globalSecondaryIndexes?: Record<
      string,
      Merge<
        GlobalSecondaryIndexProps,
        {
          partitionKey: Attr<T>;
          sortKey?: Attr<T>;
          indexName?: never;
        }
      >
    >;
  }
>;

export const createTableOptions = <T extends Record<string, any>>({
  tableName,
  ...tableProps
}: Props<T>) => {
  const defaultProps: Omit<TableProps, 'partitionKey'> = {
    tableName: `${appName}-${tableName}`,
    billingMode: BillingMode.PAY_PER_REQUEST,
    removalPolicy: RemovalPolicy.DESTROY,
  };

  return deepmerge(defaultProps, tableProps);
};
