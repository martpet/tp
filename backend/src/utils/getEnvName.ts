import { Construct } from 'constructs';

import { EnvName } from '~/types';

export const getEnvName = (scope: Construct) => {
  const envName = scope.node.tryGetContext('envName') as EnvName;

  if (!envName) {
    throw Error('envName in CDK context is required');
  }

  return envName;
};
