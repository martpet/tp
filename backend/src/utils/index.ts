export * from '~/../../shared/utils';
//
export * from './constructsUtils';
export * from './createTableOptions';
export * from './getEnvName';
export * from './getRandomBase64UrlSafe';
export * from './getRoleCredentials';
