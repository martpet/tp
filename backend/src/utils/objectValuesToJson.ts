import { JsonValue } from 'type-fest';

export const objectValuesToJson = (input: Record<string, JsonValue>) =>
  Object.fromEntries(
    Object.entries(input).map(([key, value]) => [key, JSON.stringify(value)])
  );
