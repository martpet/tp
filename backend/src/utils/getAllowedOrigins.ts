import { Construct } from 'constructs';

import { appEnvs, localhostUrl } from '~/consts';

import { getEnvName } from './getEnvName';

export const getAllowedOrigins = (scope: Construct) => {
  const envName = getEnvName(scope);
  const { appDomain } = appEnvs[envName];

  const origins = [`https://${appDomain}`];

  if (envName === 'personal') {
    origins.push(localhostUrl);
  }

  return origins;
};
