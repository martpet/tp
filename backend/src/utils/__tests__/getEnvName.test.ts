import { Construct } from 'constructs';

import { getEnvName } from '~/utils/getEnvName';

const scope = {
  node: {
    tryGetContext: () => 'dummyEnvName',
  },
} as unknown as Construct;

describe('getEnvName', () => {
  it('returns the "envName"', () => {
    expect(getEnvName(scope)).toBe('dummyEnvName');
  });

  it('throws if "envName" context is missing', () => {
    scope.node.tryGetContext = () => '';
    expect(() => {
      getEnvName(scope);
    }).toThrowErrorMatchingSnapshot();
  });
});
