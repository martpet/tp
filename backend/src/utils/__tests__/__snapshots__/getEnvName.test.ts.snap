// Vitest Snapshot v1

exports[`getEnvName > throws if "envName" context is missing 1`] = `"envName in CDK context is required"`;
